cloudify_ansible/utils.py

```diff
@@ -624,32 +742,139 @@ def get_executable_path(executable, venv):
+                str(collections_to_install),
+                collections_location))
+            install_collections_to_venv(venv_path,
+                                        collections_to_install,
+                                        collections_location)
+        else:
+            raise NonRecoverableError('No internet connection.'
+                                      'Do not use galaxy_collections when'
+                                      ' working on the plugin virtualenv.')
+
+
+def install_roles(_ctx,
+                  roles_to_install=None):
+    """
+        Handle creation of virtual environments for running playbooks.
+        The virtual environments will be created at the deployment directory.
+       :param _ctx: cloudify context.
+       :param roles_to_install: list of roles to install
+       """
+
+    if roles_to_install:
+        if is_connected_to_internet():
+            venv_path =\
+                get_instance(ctx).runtime_properties.get(PLAYBOOK_VENV)
+            roles_location = _get_roles_location(_ctx)
+            _ctx.logger.info("Installing roles {} to path {}".format(
+                str(roles_to_install),
+                roles_location))
+            install_roles_to_venv(venv_path,
+                                  roles_to_install,
+                                  roles_location)
+        else:
+            raise NonRecoverableError('No internet connection.'
+                                      'Do not use roles when'
+                                      ' working on the plugin virtualenv.')
+
+
+def install_extra_packages(_ctx,
+                           packages_to_install=None):
+    """
+        Handle creation of virtual environments for running playbooks.
+        The virtual environments will be created at the deployment directory.
+       :param _ctx: cloudify context.
+       :param packages_to_install: list of python packages to install
+        inside venv.
+       """
+
+    if packages_to_install:
+        if is_connected_to_internet():
+            if is_local_venv():
+                venv_path = \
+                    get_instance(_ctx).runtime_properties.get(PLAYBOOK_VENV)
+                _ctx.logger.info(
+                    "Installing extra_packages {} to path {}".format(
+                        str(packages_to_install),
+                        venv_path)
+                )
+                install_packages_to_venv(venv_path, packages_to_install)
+        else:
+            raise NonRecoverableError('No internet connection.'
+                                      'Do not use extra_packages when'
+                                      ' working on the plugin virtualenv.')
+
+
+def create_playbook_venv(_ctx):
+    """
+        Handle creation of virtual environments for running playbooks.
+        The virtual environments will be created at the deployment directory.
+       :param _ctx: cloudify context.
+       """
+
```

Comment:
Can we do `get_instance(_ctx)` once per function where the instance is used?

for example:

```
            instance = get_instance(_ctx)
            if is_connected_to_internet():
                ...
                instance.runtime_properties[PLAYBOOK_VENV] = venv_path
                instance.runtime_properties[LOCAL_VENV] = True
            else:
                instance.runtime_properties[PLAYBOOK_VENV] = ''
```